File: 0567-permutation-in-string/0567-permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        String pattern=s1;
        String str=s2;
        int windowStart=0;
        int matched=0;
        Map<Character,Integer> charFrequencyMap=new HashMap<>();
        for (char c : pattern.toCharArray()) charFrequencyMap.merge(c,1,Integer::sum);

        for (int windowEnd = 0; windowEnd < str.length(); windowEnd++) {
            char rightChar = str.charAt(windowEnd);
            if(charFrequencyMap.containsKey(rightChar)){
                charFrequencyMap.merge(rightChar,-1,Integer::sum);
                if(charFrequencyMap.get(rightChar)==0)matched++;
            }
            if(matched==charFrequencyMap.size())return true;
            if(windowEnd-windowStart+1==pattern.length()){
                char leftChar = str.charAt(windowStart);
                if(charFrequencyMap.containsKey(leftChar)){
                    if(charFrequencyMap.get(leftChar)==0)matched--;
                    charFrequencyMap.merge(leftChar,1,Integer::sum);
                }
                windowStart++;
            }
        }
        return false;
        
    }
}

File: 0567-permutation-in-string/0567-permutation-in-string.py
from itertools import permutations
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        n = len(s1)
        h2 = [0]*26
        h1 = [0]*26
        
        for c in s1: 
            h1[ord(c) - ord('a')] += 1
        
        for i in range(len(s2)):
            if i < n:
                h2[ord(s2[i]) - ord('a')] += 1
            else:
                h2[ord(s2[i]) - ord('a')] += 1
                h2[ord(s2[i-n]) - ord('a')] -= 1
                
            if h1 == h2: return True
            
        return False
        